### PR TITLE
Fix error message for HAPI-0302 showing misleading names for known resource types

### LINKED ISSUE #5874: Error message for HAPI-0302 shows misleading names for known resource types
**Describe the bug**
When a request is made for an unknown resource type, the server returns the resource types that it does know how to handle. When it does so, "List" is shown as "ListResource", which can (and already has) mislead someone trying to debug the error, especially if they then try to request a `ListResource` (which doesn't exist).

**To Reproduce**
1. Make a GET request to any invalid resource, like `GET /Abcdefgh/1326`.
2. See that "ListResource" is in the error message.

**Expected behavior**
The "List" resource is correctly shown as "List" in the list of resource types.

**Screenshots**
image

image

**Environment:**
 - HAPI FHIR Version: 7.3.0-SNAPSHOT

**Additional context**
Fix is ready and incoming.

I believe that this happens specifically with "List" because `List` is already a type in the Java standard library, so the resource class for "List" is named differently.

### PR DESCRIPTION
**Pull request by holly-smile**
Closes #5874.

### CODE REVIEW

**Comment by holly-smile** [comment]
Sorry, still needs a unit test - incoming!

**Review by jmarchionatto — CHANGES_REQUESTED** [review]
I am not sure about this fix. 
There is a provider returning "ListResource" name, apparently to meet some DSTU2 version spec, and we let that happen to later change it, no matter the version. 
I would rather discriminate the name by version, if necessary at all, in the proviuder, and eliminate the introduced remapping.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/RestfulServer.java [hunk 1]
```diff
@@ -1990,14 +1991,22 @@ protected void throwUnknownResourceTypeException(String theResourceName) {
...
+				.map(RestfulServer::getKnownDistinctAndSortedResourceTypes)
 				.distinct()
 				.sorted()
 				.collect(toList());
 		throw new ResourceNotFoundException(Msg.code(302) + "Unknown resource type '" + theResourceName
 				+ "' - Server knows how to handle: " + knownDistinctAndSortedResourceTypes);
 	}
 
+	private static String getKnownDistinctAndSortedResourceTypes(IResourceProvider theResourceProvider) {
+		String simpleName = theResourceProvider.getResourceType().getSimpleName();
```
Comment (jamesagnew) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/RestfulServer.java:2003]: We shouldn't need to maintain a hardcoded map for these. You can query the FhirContext for the resource name associated with a given class by calling something like:
```
theCtx.getResourceDefinition(resourceType).getName()
```

Comment (holly-smile) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/RestfulServer.java:2003]: Ah, that's so much cleaner (and less hacky)! Thank you!

**Review by jmarchionatto — APPROVED** [review]
Much better. Thanks! Approved.

**Pull request merged**